it will throw exception in my androd phone

Caused by: com.vaadin.event.ListenerMethod$MethodException: Invocation of method modifyBootstrapPage in org.vaadin.touchkit.settings.TouchKitSettings failed.
	at com.vaadin.event.ListenerMethod.receiveEvent(ListenerMethod.java:519)
	at com.vaadin.event.EventRouter.fireEvent(EventRouter.java:273)
	at com.vaadin.event.EventRouter.fireEvent(EventRouter.java:237)
	at com.vaadin.server.VaadinSession.modifyBootstrapResponse(VaadinSession.java:882)
	at com.vaadin.server.BootstrapHandler.getBootstrapHtml(BootstrapHandler.java:461)
	at com.vaadin.server.BootstrapHandler.synchronizedHandleRequest(BootstrapHandler.java:429)
	at com.vaadin.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:40)
	at com.vaadin.server.VaadinService.handleRequest(VaadinService.java:1601)
	... 45 more
Caused by: java.lang.IllegalArgumentException: Illegal character in path at index 10: deferredjs\F8B727D236F0B241566A8E7ACADF6D8D\1.cache.js
	at java.net.URI.create(URI.java:852)
	at java.net.URI.resolve(URI.java:1036)
	at org.vaadin.touchkit.settings.TouchKitSettings.modifyBootstrapPage(TouchKitSettings.java:266)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.vaadin.event.ListenerMethod.receiveEvent(ListenerMethod.java:499)
	... 52 more
Caused by: java.net.URISyntaxException: Illegal character in path at index 10: deferredjs\F8B727D236F0B241566A8E7ACADF6D8D\1.cache.js
	at java.net.URI$Parser.fail(URI.java:2848)
	at java.net.URI$Parser.checkChars(URI.java:3021)
	at java.net.URI$Parser.parseHierarchical(URI.java:3105)
	at java.net.URI$Parser.parse(URI.java:3063)
	at java.net.URI.<init>(URI.java:588)
	at java.net.URI.create(URI.java:850)

--- a/vaadin-touchkit-agpl/pom.xml
+++ b/vaadin-touchkit-agpl/pom.xml
@@ -158,7 +158,7 @@
         <license.short.name>agpl</license.short.name>
         <license.version>3.0</license.version>
         <vaadin.version.maven>[7.6.0.beta1,7.99.9999]</vaadin.version.maven>
-        <vaadin.plugin.version>7.6.0.beta1</vaadin.plugin.version>
+        <vaadin.plugin.version>7.7.26</vaadin.plugin.version>
         <gpg.passphrase.file>empty.properties</gpg.passphrase.file>
         <gpg.skip>true</gpg.skip>
         <snapshot.repository.url>http://oss.sonatype.org/content/repositories/vaadin-snapshots/</snapshot.repository.url>
@@ -168,7 +168,7 @@
         <Implementation-Title>Vaadin TouchKit</Implementation-Title>
         <Implementation-Vendor>${project.organization.name}</Implementation-Vendor>
         <Vaadin-Addon>vaadin-touchkit-${license.short.name}-${project.version}.jar</Vaadin-Addon>
-        <jetty.version>8.1.12.v20130726</jetty.version>
+        <jetty.version>9.4.31.v20200723</jetty.version>
     </properties>
 
     <distributionManagement>

--- a/vaadin-touchkit-agpl/src/main/java/com/vaadin/addon/touchkit/gwt/client/communication/TouchKitConnectionStateHandler.java
+++ b/vaadin-touchkit-agpl/src/main/java/com/vaadin/addon/touchkit/gwt/client/communication/TouchKitConnectionStateHandler.java
@@ -6,6 +6,7 @@ import static com.vaadin.addon.touchkit.gwt.client.offlinemode.OfflineMode.Activ
 import com.google.gwt.core.client.JavaScriptObject;
 import com.google.gwt.http.client.Request;
 import com.google.gwt.http.client.Response;
+import com.google.gwt.user.client.Window;
 import com.vaadin.addon.touchkit.gwt.client.offlinemode.OfflineModeEntrypoint;
 import com.vaadin.client.communication.ConnectionStateHandler;
 import com.vaadin.client.communication.DefaultConnectionStateHandler;
@@ -40,7 +41,14 @@ public class TouchKitConnectionStateHandler extends
 
     @Override
     public void heartbeatInvalidStatusCode(Request request, Response response) {
-        if (!offlineModeEntrypoint.isOfflineModeEnabled()) {
+		if(response.getStatusCode() == 410) {
+			// This has broken at some point, session expired don't properly 
+			// restart the application. Do a manual reload as I don't know how
+			// it should be done these days...
+			// TODO ask somebody who knows about this mess
+			Window.Location.reload();
+		}
+		if (!offlineModeEntrypoint.isOfflineModeEnabled()) {
             super.heartbeatInvalidStatusCode(request, response);
         } else {
             offlineModeEntrypoint.dispatch(BAD_RESPONSE);

--- a/vaadin-touchkit-agpl/src/main/java/com/vaadin/addon/touchkit/gwt/client/offlinemode/CacheManifestStatusIndicator.java
+++ b/vaadin-touchkit-agpl/src/main/java/com/vaadin/addon/touchkit/gwt/client/offlinemode/CacheManifestStatusIndicator.java
@@ -253,11 +253,15 @@ public class CacheManifestStatusIndicator implements EntryPoint {
 
     /**
      * @return The status of the application cache. See the constants in this
-     *         class for possible values.
+     *         class for possible values. Return 99 if application cache does 
+     *         not exist.
      */
     private static native int getStatus()
     /*-{
-        return $wnd.applicationCache.status;
+		if($wnd.applicationCache) {
+        	return $wnd.applicationCache.status;
+		}
+        return 99;
     }-*/;
 
     /**

--- a/vaadin-touchkit-agpl/src/main/java/com/vaadin/addon/touchkit/settings/ApplicationCacheSettings.java
+++ b/vaadin-touchkit-agpl/src/main/java/com/vaadin/addon/touchkit/settings/ApplicationCacheSettings.java
@@ -117,7 +117,7 @@ public class ApplicationCacheSettings implements BootstrapListener {
      *         enabled.
      */
     public boolean isCacheManifestEnabled() {
-        return cacheManifestEnabled;
+    	return cacheManifestEnabled && !TouchKitSettings.supportsGooglePWA();
     }
 
     /**

--- a/vaadin-touchkit-agpl/src/main/java/com/vaadin/addon/touchkit/settings/TouchKitSettings.java
+++ b/vaadin-touchkit-agpl/src/main/java/com/vaadin/addon/touchkit/settings/TouchKitSettings.java
@@ -1,23 +1,42 @@
 package com.vaadin.addon.touchkit.settings;
 
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintWriter;
 import java.io.Serializable;
+import java.net.URI;
+import java.net.URISyntaxException;
 import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.commons.io.IOUtils;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
 
 import com.vaadin.addon.touchkit.annotations.CacheManifestEnabled;
 import com.vaadin.addon.touchkit.annotations.OfflineModeEnabled;
 import com.vaadin.addon.touchkit.server.TouchKitServlet;
+import com.vaadin.addon.touchkit.service.ApplicationIcon;
 import com.vaadin.server.BootstrapFragmentResponse;
 import com.vaadin.server.BootstrapListener;
 import com.vaadin.server.BootstrapPageResponse;
 import com.vaadin.server.CustomizedSystemMessages;
+import com.vaadin.server.RequestHandler;
 import com.vaadin.server.ServiceException;
 import com.vaadin.server.SessionInitEvent;
 import com.vaadin.server.SessionInitListener;
 import com.vaadin.server.SystemMessages;
 import com.vaadin.server.SystemMessagesInfo;
 import com.vaadin.server.SystemMessagesProvider;
+import com.vaadin.server.VaadinRequest;
+import com.vaadin.server.VaadinResponse;
 import com.vaadin.server.VaadinService;
 import com.vaadin.server.VaadinServletService;
+import com.vaadin.server.VaadinSession;
 
 /**
  * TouchKitSettings is a collection of tools that help modify various touch
@@ -30,6 +49,8 @@ import com.vaadin.server.VaadinServletService;
 @SuppressWarnings("serial")
 public class TouchKitSettings implements BootstrapListener,
         SessionInitListener, SystemMessagesProvider {
+
+    private List<String> stronglyCachedResources;
 
     /**
      * Interface to select different settings for different kind of devices
@@ -185,11 +206,175 @@ public class TouchKitSettings implements BootstrapListener,
                     offline == null || offline.value());
             getApplicationCacheSettings().modifyBootstrapPage(response);
         }
+
+        if (supportsGooglePWA()) {
+
+            String contextPath = response.getRequest().getContextPath();
+
+            Document document = response.getDocument();
+            // manifest.json
+            // <link rel="manifest" href="/manifest.json">
+            Element element = document.createElement("link");
+            element.attr("rel", "manifest");
+            element.attr("href", contextPath + "/manifest.json");
+            document.getElementsByTag("head").get(0).appendChild(element);
+            // This meta tag is for some weird reason needed for 100% google PWA ;-)
+            element = document.createElement("meta");
+            element.attr("name", "theme-color");
+            element.attr("content", getWebAppSettings().getTheme_color());
+            document.getElementsByTag("head").get(0).appendChild(element);
+
+            if (stronglyCachedResources == null) {
+                // TODO make this somehow more stable and cleaner
+                String text = document.body().toString().replaceAll("\n", "");
+
+                Pattern p = Pattern.compile(".*widgetset\": \"([^\"]+)\"", Pattern.DOTALL);
+                Matcher matcher = p.matcher(text);
+                boolean find = matcher.find();
+                String wsname = matcher.group(1);
+
+                InputStream resourceAsStream = getClass().getResourceAsStream("/VAADIN/widgetsets/" + wsname + "/" + "safari.manifest");
+
+                // Might be null in case of for example fallback UI
+                if (resourceAsStream != null) {
+                    List<String> resources = new ArrayList<String>();
+
+                    try {
+                        final URI base = new URI(contextPath + "/VAADIN/widgetsets/" + wsname + "/");
+                        // the safari permutation is used by most mobile web apps
+                        // Read the cached files from the GWT generated manifest file for service workers as well
+                        List<String> readLines = IOUtils.readLines(resourceAsStream);
+                        boolean cachedFilesStarted = false;
+                        for (String readLine : readLines) {
+                            if (readLine.startsWith("CACHE:")) {
+                                cachedFilesStarted = true;
+                                resources.add(contextPath + "/");
+                                continue;
+                            }
+                            if (readLine.startsWith("NETWORK:")) {
+                                cachedFilesStarted = false;
+                                continue;
+                            }
+                            if (cachedFilesStarted) {
+                                readLine = readLine.trim();
+                                if (!readLine.isEmpty()) {
+                                	readLine = readLine.replace("\\", "/");
+                                    URI resolved = base.resolve(readLine);
+                                    resources.add(resolved.toString());
+                                }
+                            }
+                        }
+                    } catch (IOException ex) {
+                        Logger.getLogger(TouchKitSettings.class.getName()).log(Level.SEVERE, null, ex);
+                    } catch (URISyntaxException ex) {
+                        Logger.getLogger(TouchKitSettings.class.getName()).log(Level.SEVERE, null, ex);
+                    }
+                    this.stronglyCachedResources = resources;
+                }
+
+            }
+            if (stronglyCachedResources != null && !stronglyCachedResources.isEmpty()) {
+                Element serviceworkerregistration = document.createElement("script");
+                serviceworkerregistration.attr("type", "text/javascript");
+                serviceworkerregistration.appendText("if ('serviceWorker' in navigator) {\n"
+                        + "  navigator.serviceWorker.register('"+contextPath+"/service-worker.js', { scope: '"+contextPath+"/' }).then(function(reg) {\n"
+                        + "\n"
+                        + "    if(reg.installing) {\n"
+                        + "      console.log('Service worker installing');\n"
+                        + "    } else if(reg.waiting) {\n"
+                        + "      console.log('Service worker installed');\n"
+                        + "    } else if(reg.active) {\n"
+                        + "      console.log('Service worker active');\n"
+                        + "    }\n"
+                        + "\n"
+                        + "  }).catch(function(error) {\n"
+                        + "    console.log('Registration failed with ' + error);\n"
+                        + "  });\n"
+                        + "}");
+                document.body().appendChild(serviceworkerregistration);
+            }
+
+        }        
     }
 
     @Override
     public void sessionInit(SessionInitEvent event) throws ServiceException {
         event.getSession().addBootstrapListener(this);
+        event.getSession().addRequestHandler(new RequestHandler() {
+            @Override
+            public boolean handleRequest(VaadinSession session, VaadinRequest request, VaadinResponse response) throws IOException {
+                final String pathInfo = request.getPathInfo();
+                String contextPath = request.getContextPath();
+                if (pathInfo.endsWith("manifest.json")) {
+                    // TODO write manifest.json using Jackson or similar
+                    PrintWriter writer = response.getWriter();
+					writer.append("{\n"
+                            + "  \"short_name\": \"" + getWebAppSettings().getApplicationShortName() + "\",\n"
+                            + "  \"name\": \"" + getWebAppSettings().getApplicationName() + "\",\n"
+                            + "  \"display\": \"" + getWebAppSettings().getDisplay() + "\",\n"
+                            + "  \"start_url\": \"" + contextPath + getWebAppSettings().getStart_url() + "\",\n"
+                            + "  \"background_color\": \"" + getWebAppSettings().getBackground_color() + "\",\n"
+                            + "  \"theme_color\": \"" + getWebAppSettings().getTheme_color() + "\",\n"
+                            + "  \"icons\": [\n");
+
+                    final ApplicationIcon[] icons = getApplicationIcons().getApplicationIcons();
+                    for (int i = 0; i < icons.length; i++) {
+                        ApplicationIcon icon = icons[i];
+                        if (i > 0) {
+                            writer.println(",");
+                        }
+                        writer.print("    {\n");
+                        writer.print("\"src\": \"" + icon.getHref() + "\",");
+                        writer.print("\"sizes\": \"" + icon.getSizes() + "\"");
+                        writer.print("    }\n");
+                    }
+                    writer.print("    ]\n");
+                    writer.append("}");
+
+                    return true;
+                } else if (pathInfo.endsWith("service-worker.js")) {
+                    response.setContentType("text/javascript");
+                    response.setCacheTime(-1);
+                    PrintWriter writer = response.getWriter();
+                    writer.write("self.addEventListener('install', e => {\n"
+                            + "  e.waitUntil(\n"
+                            + "    caches.open(\"tkcache\").then(c => {\n"
+                            + "      return c.addAll([\n");
+
+                    // TODO consider using https://github.com/GoogleChromeLabs/sw-precache
+                    if (stronglyCachedResources != null) {
+                        for (String stronglyCachedResource : stronglyCachedResources) {
+                            writer.write("'");
+                            writer.write(stronglyCachedResource);
+                            writer.write("',\n");
+                        }
+                    } else {
+                        Logger.getLogger(TouchKitSettings.class.getName()).log(Level.SEVERE, "strongly cached resources could not be found");
+                    }
+
+                    writer.write(""
+                            + "      ]).then(() => self.skipWaiting());\n"
+                            + "    })\n"
+                            + "  );\n"
+                            + "});\n"
+                            + "\n"
+                            + "self.addEventListener('fetch', e => {\n"
+                            + "  e.respondWith(\n"
+                            + "    caches.open(\"tkcache\").then(c => {\n"
+                            + "      return c.match(e.request).then(res => {\n"
+                            + "        return res || fetch(e.request)\n"
+                            + "      });\n"
+                            + "    })\n"
+                            + "  );\n"
+                            + "});");
+                    return true;
+                }
+
+                return false;
+            }
+
+        }
+        );
     }
 
     /**
@@ -260,5 +445,23 @@ public class TouchKitSettings implements BootstrapListener,
                 .setCommunicationErrorNotificationEnabled(false);
         customizedSystemMessages.setSessionExpiredNotificationEnabled(false);
         return customizedSystemMessages;
+    }
+
+    /**
+     * @return true if Google style service worker + json manifest style PWA 
+     * should be used instead of original iOS style home screen web app thingies.
+     */
+    public static boolean supportsGooglePWA() {
+        try {
+            VaadinRequest currentRequest = VaadinServletService.getCurrentRequest();
+            String useragentheader = currentRequest.getHeader("User-Agent").toLowerCase();
+            // Simply expect all chromes to support
+            if (useragentheader.contains("chrome")) {
+                return true;
+            }
+        } catch (Exception e) {
+            System.err.println("Detecting sw support failed!");
+        }
+        return false;
     }
 }

--- a/vaadin-touchkit-agpl/src/main/java/com/vaadin/addon/touchkit/settings/WebAppSettings.java
+++ b/vaadin-touchkit-agpl/src/main/java/com/vaadin/addon/touchkit/settings/WebAppSettings.java
@@ -23,6 +23,44 @@ public class WebAppSettings implements BootstrapListener {
     private boolean webAppCapable = true;
     private String statusBarStyle = "black";
     private String startupImage;
+    private String applicationName;
+    private String applicationShortName;
+    private String display = "fullscreen";
+    private String start_url = "/";
+    private String background_color = "white";
+    private String theme_color = "gray";
+
+    public void setTheme_color(String theme_color) {
+        this.theme_color = theme_color;
+    }
+
+    public String getTheme_color() {
+        return theme_color;
+    }
+
+    public String getBackground_color() {
+        return background_color;
+    }
+
+    public void setBackground_color(String background_color) {
+        this.background_color = background_color;
+    }
+
+    public String getStart_url() {
+        return start_url;
+    }
+
+    public void setStart_url(String start_url) {
+        this.start_url = start_url;
+    }
+
+    public String getDisplay() {
+        return display;
+    }
+
+    public void setDisplay(String display) {
+        this.display = display;
+    }
 
     /**
      * Sets a header that tells the client whether the application is designed
@@ -92,6 +130,22 @@ public class WebAppSettings implements BootstrapListener {
         return startupImage;
     }
 
+    public String getApplicationName() {
+        return applicationName;
+    }
+
+    public void setApplicationName(String applicationName) {
+        this.applicationName = applicationName;
+    }
+
+    public String getApplicationShortName() {
+        return applicationShortName;
+    }
+
+    public void setApplicationShortName(String applicationShortName) {
+        this.applicationShortName = applicationShortName;
+    }
+    
     @Override
     public void modifyBootstrapFragment(BootstrapFragmentResponse response) {
         // NOP
@@ -124,6 +178,13 @@ public class WebAppSettings implements BootstrapListener {
             element.attr("rel", "apple-touch-startup-image");
             element.attr("href", getStartupImage());
             head.appendChild(element);
+        }
+        if (getApplicationShortName() != null) {
+            element = document.createElement("meta");
+            element.attr("name", "apple-mobile-web-app-title");
+            element.attr("content", getApplicationShortName());
+            head.appendChild(element);
+        		//
         }
 
         /*

--- a/vaadin-touchkit-agpl/src/test/java/com/vaadin/addon/touchkit/TestTouchKitServlet.java
+++ b/vaadin-touchkit-agpl/src/test/java/com/vaadin/addon/touchkit/TestTouchKitServlet.java
@@ -1,0 +1,24 @@
+package com.vaadin.addon.touchkit;
+
+import javax.servlet.ServletException;
+
+import com.vaadin.addon.touchkit.server.TouchKitServlet;
+import com.vaadin.addon.touchkit.settings.TouchKitSettings;
+
+/**
+ *
+ * @author mstahv
+ */
+public class TestTouchKitServlet extends TouchKitServlet {
+
+    @Override
+    protected void servletInitialized() throws ServletException {
+        super.servletInitialized();
+        TouchKitSettings touchKitSettings = getTouchKitSettings();
+
+        touchKitSettings.getWebAppSettings().setApplicationName("TouchKit test app");
+        touchKitSettings.getWebAppSettings().setApplicationShortName("TouchKitTest");
+        touchKitSettings.getApplicationIcons().addApplicationIcon(128, 128, "image.png", true);
+
+    }
+}

--- a/vaadin-touchkit-agpl/src/test/java/com/vaadin/addon/touchkit/TouchkitTestUI.java
+++ b/vaadin-touchkit-agpl/src/test/java/com/vaadin/addon/touchkit/TouchkitTestUI.java
@@ -43,6 +43,9 @@ public class TouchkitTestUI extends UI {
                     className = getClass().getPackage().getName() + ".itest."
                             + requestPathInfo.substring(1);
                 }
+                if (className.contains("/")) {
+                    className = className.substring(0, className.indexOf("/"));
+                }
                 Class<?> forName = Class.forName(className);
                 if (forName != null) {
                     AbstractComponentContainer newInstance = (AbstractComponentContainer) forName

--- a/vaadin-touchkit-cval/pom.xml
+++ b/vaadin-touchkit-cval/pom.xml
@@ -57,7 +57,7 @@
         <Implementation-Vendor>${project.organization.name}</Implementation-Vendor>
         <Vaadin-Addon>vaadin-touchkit-${license.short.name}-${project.version}.jar</Vaadin-Addon>
 
-        <jetty.version>8.1.12.v20130726</jetty.version>
+        <jetty.version>9.4.31.v20200723</jetty.version>
 
 
     </properties>

